[coordinator] Split NewConnectionFn into NewClientFn

Antanukas opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, please read our contributor guidelines: https://github.com/m3db/m3/blob/master/CONTRIBUTING.md and developer notes https://github.com/m3db/m3/blob/master/DEVELOPMENT.md
2. Please prefix the name of the pull request with the component you are updating in the format "[component] Change title" (for example "[dbnode] Support out of order writes") and also label this pull request according to what type of issue you are addressing. Furthermore, if this is a WIP or DRAFT PR, please create a draft PR instead: https://github.blog/2019-02-14-introducing-draft-pull-requests/
3. Ensure you have added or ran the appropriate tests for your PR: read more at https://github.com/m3db/m3/blob/master/DEVELOPMENT.md#testing-changes
4. Follow the instructions for writing a changelog note: read more at https://github.com/m3db/m3/blob/master/DEVELOPMENT.md#updating-the-changelog
-->

**What this PR does / why we need it**:
<!--
If you have an issue this change addresses, please add the following details:
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Refactors code to reduce probability of introducing channel leak in the future.

**Special notes for your reviewer**:

Refactoring code related to opening a TChannel and constructing rpc client. Both session and connection_pool will reuse `establishConnection`.

1. Adds `NewClientFn`, `HealthCheckNewConnFn`, `HealthCheckFn` to be able to have a finer grained configuration of connection pool.
2. `NewConnectionFn` is now responsible to create TChannel only. Client creation moved to  `NewClientFn`.
3. `establishNewConnection` private func introduced to cover full lifecycle of creating new connection and ensuring channel is closed in case of failure.
4. New connection errors are now logged with `Warn` instead of debug. It should not be spammy since pool tries to reconnect everything every 4 seconds.

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note

```

**Does this PR require updating code package or user-facing documentation?**:
<!--
If no, just write "NONE" in the documentation-note block below.
If yes, describe which documentation you updated.
Note: Any changes that significantly updates how a code package is architectured or functions requires code package documentation updates in the form of updates to the README.md file in that package.
-->
```documentation-note

```

Antanukas commented on src/dbnode/client/connection_pool.go:
```diff
@@ -182,30 +182,25 @@ func (p *connPool) connectEvery(interval time.Duration, stutter time.Duration) {
+					p.pool = append(p.pool, conn{ch, client})
 					p.poolLen = int64(len(p.pool))
+				} else {
+					ch.Close()
```
This is likely a corner case since pool is being closed on shutdown only. Just trying to be a good citizen and cleanup. Also did not figure an easy way to test this.

Antanukas commented on src/dbnode/client/options.go:
```diff
@@ -327,26 +328,6 @@ func NewOptionsForAsyncClusters(opts Options, topoInits []topology.Initializer,
 	return result
 }
 
-func defaultNewConnectionFn(
```
moved to connection_pool.go and split into 2 methods: NewConnectionFn and NewClientFn

Antanukas commented on src/dbnode/client/session.go:
```diff
@@ -777,18 +775,12 @@ func (s *session) DedicatedConnection(
 		}
 
-		if err := s.healthCheckNewConnFn(client, s.opts, opts.BootstrappedNodesOnly); err != nil {
-			multiErr = multiErr.Add(err)
```
here we were missing a channel close statement. Now this is fixed by reusing `establishNewConnection`

linasm commented:
Could we just add the missing `close` instead of a whole rewrite of critical piece of code? Or at least do it in 2 separate PRs?

Antanukas commented:
If I split to separate PRs it would require double work - I will need to implement additional test suite in `session` and in `connection_pool.go`. By doing this refactoring I moved testing to `establishConnection` and removed responsibility of closing a channel in session.go

ryanhall07 commented on src/dbnode/client/connection_pool.go:
```diff
@@ -182,34 +182,27 @@ func (p *connPool) connectEvery(interval time.Duration, stutter time.Duration) {
 				defer wg.Done()
 
-				// Create connection
-				channel, client, err := newConnFn(channelName, address, p.opts)
```
how was the channel being leaked previously?

ryanhall07 commented on src/dbnode/client/connection_pool.go:
```diff
@@ -182,34 +182,27 @@ func (p *connPool) connectEvery(interval time.Duration, stutter time.Duration) {
 				defer wg.Done()
 
-				// Create connection
-				channel, client, err := newConnFn(channelName, address, p.opts)
```
i don't follow what the original leak was

Antanukas replied:
Here there was no leak. I have extracted code into establishConnection and reused it for dedicated session that had a leak before I fixed it here https://github.com/m3db/m3/pull/4026

Antanukas commented:
I did split this PR into smaller one with channels closed https://github.com/m3db/m3/pull/4026
